File: src/swaperex/web/controllers/quotes.py
"""Quote API endpoints."""

from fastapi import APIRouter, HTTPException

from swaperex.web.contracts.quotes import (
    QuoteRequest,
    QuoteResponse,
    MultiQuoteRequest,
    MultiQuoteResponse,
)
from swaperex.web.services.quote_service import QuoteService

router = APIRouter(prefix="/quotes", tags=["quotes"])

# Service instance
_quote_service = QuoteService()


@router.post("/", response_model=QuoteResponse)
async def get_quote(request: QuoteRequest) -> QuoteResponse:
    """Get a swap quote.

    Returns the best available quote for swapping from_asset to to_asset.
    This is a READ-ONLY operation - no transactions are executed.
    """
    return await _quote_service.get_quote(request)


@router.post("/multi", response_model=MultiQuoteResponse)
async def get_multi_quote(request: MultiQuoteRequest) -> MultiQuoteResponse:
    """Get quotes from multiple providers.

    Returns quotes from all available providers, sorted by best rate.
    """
    return await _quote_service.get_multi_quote(request)


@router.get("/pairs")
async def get_supported_pairs() -> dict:
    """Get list of supported trading pairs.

    Returns:
        List of (from_asset, to_asset) pairs that can be quoted.
    """
    pairs = _quote_service.get_supported_pairs()
    return {
        "success": True,
        "pairs": [{"from": p[0], "to": p[1]} for p in pairs[:100]],  # Limit response
        "total": len(pairs),
    }

File: frontend/src/api/quotes.ts
/**
 * Quotes API - /quotes endpoints
 */

import apiClient from './client';
import type { QuoteRequest, QuoteResponse, MultiQuoteResponse } from '@/types/api';

export async function getQuote(request: QuoteRequest): Promise<QuoteResponse> {
  const response = await apiClient.post<QuoteResponse>('/quotes/', request);
  return response.data;
}

export async function getMultiQuote(request: QuoteRequest): Promise<MultiQuoteResponse> {
  const response = await apiClient.post<MultiQuoteResponse>('/quotes/multi', request);
  return response.data;
}

export async function getSupportedPairs(): Promise<{ pairs: Array<{ from: string; to: string }>; total: number }> {
  const response = await apiClient.get('/quotes/pairs');
  return response.data;
}

export const quotesApi = {
  getQuote,
  getMultiQuote,
  getSupportedPairs,
};

export default quotesApi;
